src/irr/asset/CGLSLBRDFBuiltinIncludeLoader.h

```diff
@@ -342,6 +344,25 @@ float irr_glsl_beckmann(in float a2, in float NdotH2)
     return irr_glsl_RECIPROCAL_PI * nom/denom;
 }
 
+//TODO this is taking into account beckmann NDF only, i'd like to also account for masking/shadowing term (visibility of microfacets)
+BSDFSample irr_glsl_beckmann_cos_gen_sample(in uvec2 _sample, in float _a2)
+{
+    vec2 u = vec2(_sample)/float(4294967295);
+    float phi = 2.0*irr_glsl_PI*u.x;
+    float tan2theta = u.y==1.0 ? 0.0 : (-_a2*log(1.0-u.y));
+    float cos2Theta = 1.0 / (1.0 + tan2theta);
+    float cosTheta = sqrt(cos2Theta);
+    float sinTheta = sqrt(1.0 - cos2Theta);
+    
+    BSDFSample smpl;
+    //erm... actually i'm returning half vector here, idk how could i get L with this api
+    //user should read this H (half vector/micro-surface normal) and compute L by reflecting V on H
+    //just remember that returned vector is in reflection space so first it has to be transformed to space in which light computations are done
+    smpl.L = vec3(sinTheta,sinTheta,cosTheta)*vec3(cos(phi),sin(phi),1.0);
+    smpl.probability = irr_glsl_beckmann(_a2, cos2Theta)*abs(cosTheta);//multiply by cosTheta in order for the NDF to actually be PDF (integrates to 1)
```

Comment:
actually we need to separate out the probability (and not include `probability` in BSDF sample, just `contribution`) from the contribution
```glsl
smpl.contribution = g*fr/(4.0*interaction.NdotV); // this is what you're summing with your monte carlo estimator
probability = irr_glsl_beckmann(_a2, cos2Theta);
```

We should probably have two overloads of this function, one that has an `out probability` and one that does not

Reply:
and `contribution` should be a `Spectrum`